File: code/how-rust-saves-our-asses/script6.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

type Data struct {
	People []Person
}
type Person struct {
	Name     string
	Address  *Address
	WorksFor *Organization
}
type Organization struct {
	Name    string
	Address *Address
}

type Address struct {
	AddressLocality string
	AddressRegion   string
}

func main() {
	var data = Data{}
	if err := json.NewDecoder(os.Stdin).Decode(&data); err != nil {
		panic(err)
	}
	for _, person := range data.People {
		city := "???"
		state := "???"
		works_for_name := "???"
		works_for_city := "???"
		works_for_state := "???"

		if person.Address != nil {
			city = person.Address.AddressLocality
			state = person.Address.AddressRegion
		}
		if person.WorksFor != nil {
			works_for_name = person.WorksFor.Name
			if person.WorksFor.Address != nil {
				works_for_city = person.WorksFor.Address.AddressLocality
				works_for_state = person.WorksFor.Address.AddressRegion
			}
		}

		fmt.Printf(`
%s lives in %s %s.
They work for %s in %s, %s.
`,
			person.Name,
			city,
			state,
			works_for_name,
			works_for_city,
			works_for_state,
		)
	}
}

File: code/how-rust-saves-our-asses/src/bin/script6.rs
#[macro_use]
extern crate serde_derive;
extern crate serde;
extern crate serde_json;

use std::io;

#[derive(Deserialize, Default, Debug)]
struct Data {
    pub people: Vec<Person>
}

#[derive(Deserialize, Default, Debug)]
struct Person {
    pub name: String,
    pub address: Option<Address>,
    pub worksFor: Option<Organization>
}

#[derive(Deserialize, Default, Debug)]
struct Organization {
    pub name: String,
    pub address: Option<Address>
}

#[derive(Deserialize, Default, Debug)]
struct Address {
    pub addressLocality: String,
    pub addressRegion: String
}

fn main() {
    let data: Data = serde_json::from_reader(io::stdin())
        .expect("Invalid JSON");

    for person in data.people.into_iter() {
        let mut city = String::from("???");
        let mut state = String::from("???");
        let mut works_for_name = String::from("???");
        let mut works_for_city = String::from("???");
        let mut works_for_state = String::from("???");

        if let Some(address) = person.address {
            city = address.addressLocality;
            state = address.addressRegion;
        }
        if let Some(works_for) = person.worksFor {
            works_for_name = works_for.name;
            if let Some(address) = works_for.address {
                works_for_city = address.addressLocality;
                works_for_state = address.addressRegion;
            }
        }
        println!("
{} lives in {}, {}.
They work for {} in {} {}.
            ", 
            person.name,
            city,
            state,
            works_for_name,
            works_for_city,
            works_for_state
        );
    }
}